flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -321,8 +321,17 @@ class ExprLowering {
     const auto *symbol = proc.GetSymbol();
     // TODO: non restricted specific intrinsics, needs to get runtime symbol,
     // or think of something else.
-    if (!symbol)
-      TODO();
+    if (!symbol) {
```

Comment:
The absence of a symbol seems an odd way to determine a call is to an intrinsic. Is that the contract with the front-end or is this just taking advantage of some oddity of the front-end implementation?

Reply:
This is a contract. In Fortran, you can only call something that is an intrinsic or user defined. If a evaluate::ProcedureDesignator is user defined, it has a symbol. If it is an intrinsic, then it has a `SpecificIntrinsic`. I can reverse the usage (`if (const auto *intrinsic = proc.GetSpecificIntrinsic())`, and then querry the symbol), that will be nicer.

Reply:
The latter is more robust, which is what I mean by contract.

For the sake of argument, say there was a third mode added to this mix. Since there is a subtle assumption here that there will only ever be (a) or (b), it is further assumed that !(a) must always be (b). But when there is a (c), !(a) is imprecise and can result in hard to find bugs.